crates/hyperdrive-math/src/lp.rs

```diff
@@ -353,6 +412,164 @@ impl State {
+    ///  -z_out'(x) that this function returns:
+    ///
+    ///  -F'(x) = l * -PV'(x)
+    ///         = l * (1 - net_c'(x))
+    ///         = l * (1 + z_out'(x))
+    ///         = l * (1 - derivative)
+    ///
+    ///  When the pool is net short, we can express the derivative of the
+    ///  objective function F(x) by the derivative z_in'(x) that this
+    ///  function returns:
+    ///
+    ///  -F'(x) = l * -PV'(x)
+    ///         = l * (1 - net_c'(x))
+    ///         = l * (1 - z_in'(x))
+    ///         = l * (1 - derivative)
+    ///
+    ///  With these calculations in mind, this function rounds its result
+    ///  down so that F'(x) is overestimated. Since F'(x) is in the
+    ///  denominator of Newton's method, overestimating F'(x) helps to avoid
+    ///  overshooting the optimal solution.
+    fn calculate_shares_delta_given_bonds_delta_derivative(
+        &self,
+        bond_amount: I256,
+        original_share_reserves: FixedPoint,
+        original_bond_reserves: FixedPoint,
+        original_effective_share_reserves: FixedPoint,
+        original_share_adjustment: I256,
+    ) -> Result<FixedPoint> {
+        // Calculate the bond reserves after the bond amount is applied.
+        let bond_reserves_after = if bond_amount >= I256::zero() {
+            self.bond_reserves() + bond_amount.into()
+        } else {
+            let bond_amount = FixedPoint::from(U256::try_from(-bond_amount)?);
+            if bond_amount < self.bond_reserves() {
+                self.bond_reserves() - bond_amount
+            } else {
+                return Err(eyre!("Calculating the bond reserves underflows"));
+            }
+        };
+
+        // NOTE: Round up since this is on the rhs of the final subtraction.
+        //
+        // derivative = c * (mu * z_e(x)) ** -t_s +
+        //              (y / z_e) * (y(x)) ** -t_s -
+        //              (y / z_e) * (y(x) + dy) ** -t_s
+        let effective_share_reserves = self.effective_share_reserves_safe();
+        let derivative = self.vault_share_price().div_up(
+            self.initial_vault_share_price()
+                .mul_down(effective_share_reserves)
+                .pow(self.time_stretch()),
+        ) + original_bond_reserves.div_up(
+            original_effective_share_reserves
+                .mul_down(self.bond_reserves().pow(self.time_stretch())),
+        );
+
+        // NOTE: Rounding this down rounds the subtraction up.
+        let rhs = original_bond_reserves.div_down(
+            original_effective_share_reserves.mul_up(bond_reserves_after.pow(self.time_stretch())),
+        );
+        if derivative < rhs {
+            return Err(eyre!("Derivative is less than right hand side"));
+        }
+        let derivative = derivative - rhs;
+
+        // NOTE: Round up since this is on the rhs of the final subtraction.
+        //
+        // inner = (
+        //             (mu / c) * (k(x) - (y(x) + dy) ** (1 - t_s))
```

Comment:
this is not `k(x)` right? `k` is not dependent on `x`

Reply:
The comments are copied from the solidity.  I'd like to keep them the same, if there are things we'd like to change, then matching PRs should go up to fix both here and in solidity.

Reply:
roger; I commented in the issue mentioned above to look into this.